### PR TITLE
scheduling: sync schedulers from the API server

### PR DESCRIPTION
**Pull request by JmPotato**
### What problem does this PR solve?

Issue Number: ref #5839. Should merge after #7082.

### What is changed and how does it work?

```commit-message
- Sync the schedulers from the API server.
- Dynamically reload the scheduler config.
```

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/mcs/scheduling/server/cluster.go [hunk 1]
```diff
@@ -170,6 +176,85 @@ func (c *Cluster) SwitchAPIServerLeader(new pdpb.PDClient) bool {
 	return c.apiServerLeader.CompareAndSwap(old, new)
 }
 
+// UpdateScheduler listens on the schedulers updating notifier and manage the scheduler creation and deletion.
+func (c *Cluster) UpdateScheduler() {
+	defer logutil.LogPanic()
+
+	notifier := make(chan struct{}, 1)
+	c.persistConfig.SetSchedulersUpdatingNotifier(notifier)
```
Comment (rleungx) [inline pkg/mcs/scheduling/server/cluster.go:184]: If schedulers are initialized and we add/remove a scheduler immediately before setting the notifier, will the notifier lose the update?

File: pkg/schedule/coordinator.go [hunk 2]
```diff
@@ -481,6 +499,8 @@ func (c *Coordinator) InitSchedulers(needRun bool) {
 	if err := c.cluster.GetSchedulerConfig().Persist(c.cluster.GetStorage()); err != nil {
 		log.Error("cannot persist schedule config", errs.ZapError(err))
 	}
+
+	c.markSchedulersInitialized()
```
Comment (rleungx) [inline pkg/schedule/coordinator.go:503]: Is it necessary for PD mode?

Comment (JmPotato) [reply pkg/schedule/coordinator.go:503]: It's only used in the scheduling service, but I think leaving the mark here is okay, which is not an expensive operation.

File: pkg/mcs/scheduling/server/cluster.go (hunk 1 above)
Comment (JmPotato) [reply pkg/mcs/scheduling/server/cluster.go]: Fixed. PTAL.

File: tests/integrations/mcs/scheduling/server_test.go [hunk 3]
```diff
@@ -115,3 +119,30 @@ func (suite *serverTestSuite) TestPrimaryChange() {
...
+	re := suite.Require()
+	tc, err := tests.NewTestSchedulingCluster(suite.ctx, 1, suite.backendEndpoints)
+	re.NoError(err)
+	defer tc.Destroy()
+	tc.WaitForPrimaryServing(re)
+	schedulersController := tc.GetPrimaryServer().GetCluster().GetCoordinator().GetSchedulersController()
+	re.Len(schedulersController.GetSchedulerNames(), 5)
+	re.Nil(schedulersController.GetScheduler(schedulers.EvictLeaderName))
+	// Add a new scheduler through the API server.
+	err = suite.pdLeader.GetServer().GetHandler().AddEvictLeaderScheduler(1)
```
Comment (rleungx) [inline tests/integrations/mcs/scheduling/server_test.go:133]: How about adding the evict leader for two stores?

File: pkg/schedule/schedulers/grant_leader.go [hunk 4]
```diff
@@ -178,6 +178,24 @@ func (s *grantLeaderScheduler) EncodeConfig() ([]byte, error) {
 	return EncodeConfig(s.conf)
 }
 
+func (s *grantLeaderScheduler) ReloadConfig() error {
```
Comment (lhy1024) [inline pkg/schedule/schedulers/grant_leader.go:181]: Can we reduce the duplication in ReloadConfig?

File: pkg/schedule/coordinator.go [hunk 5]
```diff
@@ -91,19 +94,34 @@ func NewCoordinator(ctx context.Context, cluster sche.ClusterInformer, hbStreams
...
+
+// markSchedulersInitialized marks the scheduler initialization is finished.
+func (c *Coordinator) markSchedulersInitialized() {
+	c.Lock()
+	defer c.Unlock()
+	c.schedulersInitialized = true
+}
+
+// CheckSchedulersInitialized returns whether the schedulers have been initialized.
+func (c *Coordinator) CheckSchedulersInitialized() bool {
```
Comment (lhy1024) [inline pkg/schedule/coordinator.go:121]: How about `AreSchedulersInitialized`?

File: pkg/schedule/schedulers/grant_leader.go (hunk 4 above)
Comment (JmPotato) [reply pkg/schedule/schedulers/grant_leader.go:181]: Different config structures need to have different implementation, I think it's OK to have some duplication here.

**Review by rleungx — APPROVED** [review]

**Review by lhy1024 — APPROVED** [review]

**Review by lhy1024 — COMMENTED** [review]
TestSchedulerSync is failed

**Pull request merged**